src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala

```diff
@@ -446,84 +515,54 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
+        val orderedVerts =
+          if (face.normalVec(2) < -NumericUtils.EPSILON) {
+              // If z-component of normal vector points in negative z-direction, orientation
+            // needs to be reversed otherwise points will be ordered clockwise
+            rotatedVerts.sortWith(Block.ccwCompare(_, _, center)).reverse
           }
-        } else {
-          Delaunay.Triangulation(rotatedVertices).map { vert =>
-            Array(vert._1, vert._2, vert._3)
+          else {
+            rotatedVerts.sortWith(Block.ccwCompare(_, _, center))
           }
+        // Rotate vertices back to original orientation
+        val invR = R.t // Inverse of rotation matrix is equal to its transpose
+        orderedVerts map { vertex =>
+          val orderedVertex = (invR * DenseVector[Double](vertex)).map {
+            NumericUtils.roundToTolerance(_) }
+          Array(orderedVertex(0), orderedVertex(1), orderedVertex(2))
         }
       }
     ).toMap
+  }
 
   /**
-    * Calculates the centroid of the block.
-    * See http://wwwf.imperial.ac.uk/~rn/centroid.pdf for theoretical background.
+    * Calculates the centroid of the block using simplex integration.
     *
     * @return The centroid of the block, (centerX, centerY, centerZ).
     */
   def centroid: Array[Double] = {
-    val vertices = findVertices
-    val mesh = meshFaces(vertices)
-
-    val increments = faces.flatMap { face =>
-      mesh(face).map { m =>
-        // Access triangulation entries backwards so they are in counterclockwise order
-        val a_vals = vertices(face)(m(2))
-        val b_vals = vertices(face)(m(1))
-        val c_vals = vertices(face)(m(0))
-        val a = DenseVector[Double](a_vals(0), a_vals(1), a_vals(2))
-        val b = DenseVector[Double](b_vals(0), b_vals(1), b_vals(2))
-        val c = DenseVector[Double](c_vals(0), c_vals(1), c_vals(2))
-
-        val n = linalg.cross(b - a, c - a)
-        val volIncrement = a dot n
-
-        val basisVectors = Vector(
-          DenseVector[Double](1.0, 0.0, 0.0),
-          DenseVector[Double](0.0, 1.0, 0.0),
-          DenseVector[Double](0.0, 0.0, 1.0)
-        )
-        val centroidIncrement = basisVectors map { vec =>
-          1 / 24.0 * ((n dot vec) * (
-            math.pow((a + b) dot vec, 2) +
-            math.pow((b + c) dot vec, 2) +
-            math.pow((c + a) dot vec, 2)
-          ))
-        }
-        (volIncrement, (centroidIncrement(0), centroidIncrement(1), centroidIncrement(2)))
-      }
-    }
-
-    val (totalVolume, (centroidX, centroidY, centroidZ)) = increments.fold (0.0, (0.0,0.0,0.0)) {
-      case ((volInc1, (centX1, centY1, centZ1)),  (volInc2, (centX2, centY2, centZ2))) =>
-        (volInc1 + volInc2, (centX1 + centX2, centY1 + centY2, centZ1 + centZ2))
-    }
+    val vertices = orientedVertices
 
     // Check if block is extremely small
-    if (totalVolume <= NumericUtils.EPSILON) {
+    if (volume <= NumericUtils.EPSILON) {
```

Comment:
Let's clean up the logic a bit here. I think you can write something like `allVertices.map(_(0)).sum` instead of `allVertices.map { v => v(0) }` and similarly for the other sums.